src/cmp.rs

```diff
@@ -0,0 +1,1188 @@
+//     1 means some differences were found,
+//     and 2 means trouble.
+pub fn main(opts: Peekable<ArgsOs>) -> ExitCode {
+    let params = match parse_params(opts) {
+        Ok(param) => param,
+        Err(e) => {
+            eprintln!("{e}");
+            return ExitCode::from(2);
+        }
+    };
+
+    if params.from == "-" && params.to == "-"
+        || same_file::is_same_file(&params.from, &params.to).unwrap_or(false)
+    {
+        return ExitCode::SUCCESS;
+    }
+
+    // If the files have different sizes, we already know they are not identical. If we have not
+    // been asked to show even the first difference, we can quit early.
+    if params.quiet {
+        if let (Ok(a_meta), Ok(b_meta)) = (fs::metadata(&params.from), fs::metadata(&params.to)) {
+            #[cfg(not(target_os = "windows"))]
+            if a_meta.size() != b_meta.size() {
+                return ExitCode::from(1);
+            }
+            #[cfg(target_os = "windows")]
+            if a_meta.file_size() != b_meta.file_size() {
+                return ExitCode::from(1);
+            }
+        }
+    }
+
+    match cmp(&params) {
+        Ok(Cmp::Equal) => ExitCode::SUCCESS,
+        Ok(Cmp::Different) => ExitCode::from(1),
+        Err(e) => {
+            if !params.quiet {
+                eprintln!("{e}");
+            }
+            ExitCode::from(2)
+        }
+    }
+}
+
+#[inline]
+fn is_ascii_printable(byte: u8) -> bool {
+    let c = byte as char;
+    c.is_ascii() && !c.is_ascii_control()
+}
+
+#[inline]
+fn format_octal(byte: u8, buf: &mut [u8; 3]) -> &str {
+    *buf = [b' ', b' ', b'0'];
+
+    let mut num = byte;
+    let mut idx = 2; // Start at the last position in the buffer
+
+    // Generate octal digits
+    while num > 0 {
+        buf[idx] = b'0' + num % 8;
+        num /= 8;
+        idx = idx.saturating_sub(1);
+    }
+
+    // SAFETY: the operations we do above always land within ascii range.
+    unsafe { std::str::from_utf8_unchecked(&buf[..]) }
+}
+
+#[inline]
+fn format_byte(byte: u8) -> String {
+    let mut byte = byte;
+    let mut quoted = vec![];
+
+    if !is_ascii_printable(byte) {
+        if byte >= 128 {
+            quoted.push(b'M');
+            quoted.push(b'-');
+            byte -= 128;
+        }
+
+        if byte < 32 {
+            quoted.push(b'^');
+            byte += 64;
+        } else if byte == 127 {
+            quoted.push(b'^');
+            byte = b'?';
+        }
+        assert!((byte as char).is_ascii());
+    }
+
+    quoted.push(byte);
+
+    // SAFETY: the checks and shifts we do above match what cat and GNU
+    // cmp do to ensure characters fall inside the ascii range.
+    unsafe { String::from_utf8_unchecked(quoted) }
+}
+
+// This function has been optimized to not use the Rust fmt system, which
+// leads to a massive speed up when processing large files: cuts the time
+// for comparing 2 ~36MB completely different files in half on an M1 Max.
```

Comment:
I haven't run the benchmarks myself, I'll take your word for it. Out of curiosity, have you run the same benchmarks on medium-sized and small files?